### PR TITLE
Implement promised requirements (resolves #729)

### LINKED ISSUE #729: Fix Promises return value
Our group would like to broaden the scope of promises (e.g. return both a fileStoreID and file-size that can be resolved and used in the same function).

### PR DESCRIPTION
**Pull request by jpfeil**
Resolves #729, continued from #791

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
Add a pinch of dill …

… to the list of requirements in setup.py

**Comment by jpfeil** [comment]
Jenkins, test this please

**Comment by hannes-ucsc** [comment]
Merged https://github.com/BD2KGenomics/cgcloud/pull/167 and created and registered a new toil-jenkins-slave image.

Jenkins, test this please.

Fingers crossed.

**Comment by hannes-ucsc** [comment]
Master was bjørked. Retrying …

Jenkins, test this please.

**Comment by jpfeil** [comment]
@hannes-ucsc

File: src/toil/job.py [hunk 1]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
...
+
+    def get_value(self):
+        """
+        Returns PromisedRequirement value
+        """
+        func = dill.loads(self._func)
+        return func(*self._args)
+
+
+def convertPromiseToPromisedRequirement(kwargs):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Combine these two methods. It should be possible to convert any promised requirements and then return True if any conversions were made.

Also make the resulting function a static method of the PromisedRequirement class and rename it to just `convertPromises`.

File: src/toil/job.py [hunk 2]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
...
+        if hasattr(args[0], '__call__'):
+            func = args[0]
+            args = args[1:]
+        else:
+            func = lambda x: x
+
+        self._func = dill.dumps(func)
+        self._args = list(args)
+
+    def get_value(self):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Camel case

File: src/toil/job.py [hunk 3]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
             # corrupted
             value = cPickle.load(fileHandle)
             return value
+
+
+class PromisedRequirement(object):
+    def __init__(self, *args):
+        """
+        Class for manipulating Promise instances.
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Needs more meat.

File: src/toil/job.py [hunk 4]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
...
+        (see toil.job.Promise)
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Use the Sphinx syntax for class references.

File: src/toil/job.py [hunk 5]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
...
+
+        :param args[0]: function to act on instance args
```
Comment (hannes-ucsc) [inline src/toil/job.py]: args[0] won't work in Sphinx. Can you change the signature to 

```
def __init__(valueOrCallable, *args):
```

?

File: src/toil/job.py [hunk 6]
```diff
@@ -1698,3 +1783,59 @@ def _resolve(cls, jobStoreString, jobStoreFileID):
...
+                        default: lambda x: x
+        :param int|Promise args[1:]: list of args, ordered with respect to function parameters
+        """
+        if hasattr(args[0], '__call__'):
+            func = args[0]
+            args = args[1:]
+        else:
+            func = lambda x: x
```
Comment (hannes-ucsc) [inline src/toil/job.py]: What happens if I call PromisedRequirement(1,2)? Is that a legal way to call the constructor?

File: src/toil/job.py [hunk 7]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+
+        super(PromisedRequirementFunctionWrappingJob, self).__init__(userFunction, *args, **kwargs)
+
+    def run(self, fileStore):
+        # Assumes promises are fulfilled when parent job is run
+        self.replacePromisedRequirements()
+        userFunction = self._getUserFunction()
+        return self.addChildFn(userFunction, *self._args, **self._kwargs).rv()
+
+    def replacePromisedRequirements(self):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Rename to `evaluatePromisedRequirements`.

File: src/toil/job.py [hunk 8]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+class PromisedRequirementFunctionWrappingJob(FunctionWrappingJob):
+    """
+    Adds child function using parent function parameters and fulfilled Promise values.
+
+    (see :class:`toil.job.FunctionWrappingJob` and :class:`toil.job.Job`)
+    """
+
+    def __init__(self, userFunction, *args, **kwargs):
+
+        self.promised_requirement = {}
```
Comment (hannes-ucsc) [inline src/toil/job.py]: promised_requirements

Comment (hannes-ucsc) [inline src/toil/job.py]: camel case

File: src/toil/job.py [hunk 9]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+
+        # Reset any PromisedRequirement instances to default values until the promise can be fulfilled.
+        # See toil.job.PromisedRequirement
```
Comment (hannes-ucsc) [inline src/toil/job.py]: this line is redundant

File: src/toil/job.py [hunk 10]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+        requirements = {'disk': '1M', 'memory': '32M', 'cores': 0.1}
+        for requirement, default in requirements.items():
```
Comment (hannes-ucsc) [inline src/toil/job.py]: s/requirement/name and s/default/value

File: src/toil/job.py [hunk 11]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+            if requirement in kwargs and isinstance(kwargs[requirement], PromisedRequirement):
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Replace `in` followed by use of `[]` operator with `try: [] except KeyError: pass else ...`

File: src/toil/job.py [hunk 12]
```diff
@@ -1496,6 +1531,56 @@ def run(self, fileStore):
...
+    Adds child function using parent function parameters and fulfilled Promise values.
+
+    (see :class:`toil.job.FunctionWrappingJob` and :class:`toil.job.Job`)
+    """
+
+    def __init__(self, userFunction, *args, **kwargs):
+
+        self.promised_requirement = {}
+
+        # Reset any PromisedRequirement instances to default values until the promise can be fulfilled.
```
Comment (hannes-ucsc) [inline src/toil/job.py]: Don't like the use of the word default here. These are the requirements needed for the execution of the intermediate job. There is nothing "default" about this.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 13]
```diff
@@ -329,7 +329,7 @@ def resourceOffers(self, driver, offers):
             runnableTasks = []
             # TODO: In an offer, can there ever be more than one resource with the same name?
             offerCores, offerMemory, offerDisk = self._determineOfferResources(offer)
-            log.debug('Received offer %s with %i MiB memory, %i core(s) and %i MiB of disk.',
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: This should be committed separately.

File: src/toil/batchSystems/mesos/batchSystem.py [hunk 14]
```diff
@@ -329,7 +329,7 @@ def resourceOffers(self, driver, offers):
...
+            log.debug('Received offer %s with %.2f MiB memory, %.2f core(s) and %.2f MiB of disk.',
                       offer.id.value, offerMemory, offerCores, offerDisk)
```
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: This should be committed separately.

File: src/toil/leader.py [hunk 15]
```diff
@@ -151,7 +151,7 @@ def issueJob(self, jobStoreID, memory, cores, disk, preemptable):
         jobBatchSystemID = self.batchSystem.issueBatchJob(jobCommand, memory, cores, disk, preemptable)
         self.jobBatchSystemIDToIssuedJob[jobBatchSystemID] = IssuedJob(jobStoreID, memory, cores, disk, preemptable)
         logger.debug("Issued job with job store ID: %s and job batch system ID: "
-                     "%s and cores: %i, disk: %i, and memory: %i",
+                     "%s and cores: %.2f, disk: %.2f, and memory: %.2f",
```
Comment (hannes-ucsc) [inline src/toil/leader.py]: Commit separately together with similar change above.

File: src/toil/test/src/promisedRequirementTest.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+from toil.job import Job
+from toil.job import PromisedRequirement
+from toil.test import needs_mesos
+from toil.test.batchSystems.batchSystemTest import hidden
+from toil.batchSystems.mesos.test import MesosTestSupport
+
+log = logging.getLogger(__name__)
+
+
+class concealed:
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: Any reason we can't stick with the already established `hidden`

File: src/toil/test/src/promisedRequirementTest.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+    class AbstractPromisedRequirementsTest(hidden.AbstractBatchSystemJobTest):
+        """
+        An abstract base class for testing Toil workflows with promised requirements.
+        """
+
+        def testPromisedRequirementDynamic(self):
+            """
+            Asserts that promised core resources are allocated properly using a dynamic Toil workflow
+            """
+            for cores_per_job in self.allocated_cores:
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+                temp_dir = self._createTempDir('testFiles')
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+                counter_path = self.getCounterPath(temp_dir)
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+
+                root = Job.wrapJobFn(maxConcurrency, self.cpu_count, counter_path, cores_per_job,
+                                     cores=1, memory='1M', disk='1M')
+                values = Job.Runner.startToil(root, self.getOptions(temp_dir))
+                max_value = max(values)
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+            """
+            Asserts that promised core resources are allocated properly using a static DAG
+            """
+            for cores_per_job in self.allocated_cores:
+                temp_dir = self._createTempDir('testFiles')
+                counter_path = self.getCounterPath(temp_dir)
+
+                root = Job()
+                one = Job.wrapFn(getOne, cores=0.1, memory='32M', disk='1M')
+                thirty_two_mb = Job.wrapFn(getThirtyTwoMb, cores=0.1, memory='32M', disk='1M')
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+                root.addChild(one)
+                root.addChild(thirty_two_mb)
+                for _ in range(self.cpu_count):
+                    root.addFollowOn(Job.wrapFn(batchSystemTest.measureConcurrency, counter_path,
+                                                cores=PromisedRequirement(lambda x: x * cores_per_job, one.rv()),
+                                                memory=PromisedRequirement(thirty_two_mb.rv()),
+                                                disk='1M'))
+                Job.Runner.startToil(root, self.getOptions(temp_dir))
+                min_value, max_value = batchSystemTest.getCounters(counter_path)
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+
+        def getCounterPath(self, temp_dir):
+            """
+            Returns path to a counter file
+            :param str temp_dir: path to test directory
+            :return: path to counter file
+            """
+            counter_path = os.path.join(temp_dir, 'counter')
+            batchSystemTest.resetCounters(counter_path)
+            min_value, max_value = batchSystemTest.getCounters(counter_path)
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+            options.maxCores = self.cpu_count
+            return options
+
+        def getCounterPath(self, temp_dir):
+            """
+            Returns path to a counter file
+            :param str temp_dir: path to test directory
+            :return: path to counter file
+            """
+            counter_path = os.path.join(temp_dir, 'counter')
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

File: src/toil/test/src/promisedRequirementTest.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+    Returns the max number of concurrent tasks when using a PromisedRequirement instance
+    to allocate the number of cores per job.
+
+    :param int cpu_count: number of available cpus
+    :param str filename: path to counter file
+    :param int cores_per_job: number of cores assigned to each job
+    :return int max concurrency value:
+    """
+    one = job.addChildFn(getOne, cores=0.1, memory='32M', disk='1M')
+    thirty_two_mb = job.addChildFn(getThirtyTwoMb, cores=0.1, memory='32M', disk='1M')
```
Comment (hannes-ucsc) [inline src/toil/test/src/promisedRequirementTest.py]: camel case

**Comment by jpfeil** [comment]
@hannes-ucsc

**Comment by benedictpaten** [comment]
@jpfeil please address the issues regarding comments and docs we discussed today and I think we can merge this.

**Comment by hannes-ucsc** [comment]
@benedictpaten, I've set the `needs work` label to reflect your previous statement

**Comment by hannes-ucsc** [comment]
@jpfeil, @benedictpaten : Did you guys agree to add a Sphinx document section?

**Comment by jpfeil** [comment]
@hannes-ucsc Yes

**Comment by jpfeil** [comment]
Jenkins, test this please.

**Comment by jpfeil** [comment]
Jenkins, test this please.

**Comment by jpfeil** [comment]
Jenkins, test this please.

**Comment by jpfeil** [comment]
Jenkins, test this please.

**Comment by jpfeil** [comment]
@hannes-ucsc

**Comment by hannes-ucsc** [comment]
This has conflicts, I can't review until those are resolved.

**Comment by jpfeil** [comment]
Jenkins, test this please.

**Comment by hannes-ucsc** [comment]
Looks like the tests failures are directly related to this PR. Leaving as `needs work`.

**Comment by jpfeil** [comment]
@hannes-ucsc

File: src/toil/batchSystems/mesos/batchSystem.py (hunk 14 above)
Comment (hannes-ucsc) [inline src/toil/batchSystems/mesos/batchSystem.py]: Resolved.

File: setup.py [hunk 26]
```diff
@@ -25,7 +25,12 @@
     author_email='benedict@soe.usc.edu',
     url="https://github.com/BD2KGenomics/toil",
     install_requires=[
-        'bd2k-python-lib==1.13.dev14'],
+        'bd2k-python-lib==1.13.dev14',
+        'dill==0.2.5'],
+    tests_require=[
```
Comment (hannes-ucsc) [inline setup.py]: BTW, I didn't catch this earlier but this is a merge error on your part, @jpfeil. You reintroduced lines that I had taken out. Try to be extra careful with merges!

Comment (hannes-ucsc) [inline setup.py]: … sorry, with resolving merge conflicts.

**Pull request merged**